src/fsharp/service/service.fs

```diff
@@ -444,18 +444,17 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
         }
 
     member _.GetCachedCheckFileResult(builder: IncrementalBuilder, filename, sourceText: ISourceText, options) =
+        // This is to check if the check results are still valid by their timestamp.
+        // A return of None means the file needs to be re-evaluated as a dependent has changed; therefore, do not get the cached result.
+        match builder.TryGetCheckResultsForFileInProject filename with
```

Comment:
Will and I talked this through - this is basically sound and represents a fix, but two problems:

1. Better if this is just a fix to the existing `AreCheckResultsBeforeFileInProjectReady` - we only care about the BeforeFileInProject state - this will be a marginal perf improvement hitting the cache more often.

2. This requires timestamp checks on potentially 1000s of files in large solutions.  Currently we do this when caching fails, but never in the process of looking up the cache (which happens very very often, even for basic editing like quick info, when the solution is not changing at all).

For (2) an alternative is to pass in the "GetDependentVersion" from the Roslyn project to act as a first quick positive test for the validity of the cache:

```
            let cacheEntry = lookup cache
             if (project1.GetDependentVersion == cacheEntry.SavedProjectDependentVersion) 
            THEN use cacheEntry
            ELSE 
               IF AreCheckResultsBeforeFileInProjectReady... &&  sourceText.GetHashCode() = cacheEntry.SavedHashCode THEN use cacheEntry
               ELSE
                   recompute
```

Reply:
@dsyme What would the alternative look like for non-Roslyn scenarios? Can this check be a parameter lambda passed to the checker?

Reply:
@dsyme I've checked and we can compute timestamp for a project efficiently. @TIHan would it be possible to have a parameter with type like `FSharpProjectOptions -> DateTime`?

Reply:
I'm not sure what the request is here.  You can pass in a hash of the state of the world using the changes in this PR if you wish.  If not you will get the old behaviour

SO I think what's in is ok